src/stores/redis.rs

```diff
@@ -0,0 +1,436 @@
+        self.clean_up();
+        match self.base_get(&key) {
+            Some(val) => {
+                self.hits += 1;
+                self.base_vec_fill(&val)
+            }
+            None => {
+                self.misses += 1;
+                let val = f();
+                self.base_set(&key, &val);
+                let index = self.store.len();
+                self.store.push(val);
+                &mut self.store[index]
+            }
+        }
+    }
+
+    fn cache_remove(&mut self, key: &K) -> Option<V> {
+        self.clean_up();
+        let mut con = self.client.get_connection().unwrap();
+        self.base_get(&key).map(|val| {
+            con.del::<String, ()>(self.generate_key(key)).unwrap();
+            serde_json::from_str(&val).unwrap()
+        })
+    }
+
+    fn cache_clear(&mut self) {
+        self.clean_up();
+        let mut con = self.client.get_connection().unwrap();
+        redis_rs::cmd("EVAL")
+            .arg("local keys = redis.call('keys', ARGV[1]) \n for i=1,#keys,5000 do \n redis.call('del', unpack(keys, i, math.min(i+4999, #keys))) \n end \n return keys")
+            .arg(0)
+            .arg(format!("{}*", self.generate_prefix()))
+            .query(&mut con)
+            .unwrap()
+    }
+
+    /// In `RedisCache`, it's an alias to `cache_clear`
+    fn cache_reset(&mut self) {
+        self.cache_clear();
+    }
+
+    fn cache_size(&self) -> usize {
+        let mut con = self.client.get_connection().unwrap();
+        redis_rs::cmd("EVAL")
+            .arg(format!(
+                "return #redis.call('keys', '{}*')",
+                self.generate_prefix()
+            ))
+            .arg(0)
+            .query(&mut con)
+            .unwrap()
+    }
+
+    fn cache_hits(&self) -> Option<u64> {
+        Some(self.hits)
+    }
+
+    fn cache_misses(&self) -> Option<u64> {
+        Some(self.misses)
+    }
+
+    fn cache_lifespan(&self) -> Option<u64> {
+        self.seconds
+    }
+
+    fn cache_set_lifespan(&mut self, seconds: u64) -> Option<u64> {
+        self.clean_up();
+        let old = self.seconds;
+        self.seconds = Some(seconds);
+        old
+    }
+}
+
+#[cfg(feature = "async")]
+#[async_trait]
+impl<K, V> CachedAsync<K, V> for RedisCache<K, V>
+where
+    K: Send + Display,
+    V: Serialize + DeserializeOwned + Clone,
+{
+    async fn get_or_set_with<F, Fut>(&mut self, key: K, f: F) -> &mut V
+    where
+        V: Send,
+        F: FnOnce() -> Fut + Send,
+        Fut: Future<Output = V> + Send,
+    {
+        let mut con = self.client.get_connection().unwrap();
+        let generated_key = self.generate_key(&key);
+        let val: RedisResult<String> = con.get(generated_key.clone());
+        match val {
+            Ok(val) => {
+                self.hits += 1;
+                self.base_vec_fill(&val)
+            }
+            Err(_) => {
+                let val = f().await;
+                self.base_set(&key, &val);
```

Comment:
can you make the set and get interaction with redis async?

Reply:
There is a problem here.

In the redis-rs library, we have two different methods, `get_connection` which is sync and `get_async_connection` which is async. The second one will be available if you have the Cargo.toml async feature enabled. There is no problem here :) (ref: [1](https://github.com/mitsuhiko/redis-rs/blob/c005735470e487b0029d4c544867dc4ced670b4a/src/client.rs#L69))

We have some other functions for general Redis commands, like `get`, which will be sync by default, but will be async based on the trait you `use`. So same method, but different behaviors. (refs: [1](https://github.com/mitsuhiko/redis-rs/blob/c005735470e487b0029d4c544867dc4ced670b4a/src/commands.rs#L58), [2](https://github.com/mitsuhiko/redis-rs/blob/c005735470e487b0029d4c544867dc4ced670b4a/src/commands.rs#L176))

In another case, I think this Redis implementation should be all async.

But the problem is that there are two sets of `cached` methods, `Cached` and `CachedAsync` traits.
I still don't know why do we have two? Why can we disable `CachedAsync` by a feature flag and not to have it all the times? And cannot we merge them?

If it is possible to merge them and make both `async`, I think all my issues here will be solved.